### ISSUE TITLE
[python-package] [bug] Segfault when combining `use_quantized_grad` with categorical features

### ISSUE THREAD

**Issue by tony-theorem** [microsoft/LightGBM#6257]
## Description
LightGBM encounters a segfault when quantized gradients are used in combination with native categorical features.

## Reproducible examples
Training API Example:
```
import numpy as np
import lightgbm as lgbm

rng = np.random.default_rng(1)
X = rng.choice([1, 2], size=(10_000, 1))
y = rng.choice([0, 1], size=(10_000,))

train_set = lgbm.Dataset(
    data=X,
    feature_name=["A"],
    categorical_feature=["A"],
    label=y,
    params={"random_seed": 1},
    free_raw_data=False,
).construct()

booster = lgbm.train(
    params={
        "objective": "binary",
        "use_quantized_grad": True,
        "random_state": 1,
    },
    num_boost_round=10,
    train_set=train_set,
)
```

Scikit-learn API Example:
```
import numpy as np
import lightgbm as lgbm

rng = np.random.default_rng(1)
X = rng.choice([1, 2], size=(10_000, 1))
y = rng.choice([0, 1], size=(10_000,))

clf = lgbm.LGBMClassifier(
    objective="binary",
    use_quantized_grad=True,
    random_state=1,
    num_boost_round=10,
)
clf.fit(
    X=X,
    y=y,
    feature_name=["A"],
    categorical_feature=["A"],
)
```

Both examples result in a core dump when LightGBM attempts to fit the booster. The failure occurs for both regression and classification objectives. The failure also occurs regardless of the number of features as long as at least one feature is categorical. No error is encountered when either `use_quantized_grad=False` or no categorical features are specified.

## Environment info

OS: Ubuntu 22.04.3 LTS (Jammy Jellyfish)
Python Version: 3.10.8
Relevant python packages:
```
lightgbm==4.2.0
numpy==1.26.0
scikit-learn==1.3.2
scipy==1.11.4
```

`lightgbm` was built from source via `pip install --no-binary lightgbm lightgbm` using GCC=11.4.0. I also attempted to install the pre-compiled version via PyPI (`pip install lightgbm`) and encountered the same segfault.

**Comment by Innixma** [comment]
+1, I made a Colab Notebook that reproduces the issue and highlights that ordered categoricals will avoid the segfault (although this isn't an ideal workaround): https://colab.research.google.com/drive/18PCq4PAF4eWuyPSdLpWhITa_CVDArCun?usp=sharing

This bug is currently a blocker for AutoGluon to adopt the quantization logic, since we pass in categorical features to LightGBM.

**Comment by shiyu1994** [comment]
@tony-theorem @Innixma  Thanks for reporting the issue. Will investigate this.
